scripts/check_backend_associated_test_file.py

```diff
@@ -0,0 +1,134 @@
+    'core/tests/data/failing_tests.py',
+    'core/tests/data/image_constants.py',
+    'core/tests/data/unicode_and_str_handler.py',
+    'proto_files/text_classifier_pb2.py',
+    'proto_files/training_job_response_payload_pb2.py',
+
+    # The files below are to be tested.
+    # GSoC'22: Improve line and branch coverage for the backend and frontend.
+    'scripts/run_frontend_tests.py',
+    'scripts/create_expression_parser.py',
+    'scripts/run_custom_eslint_tests.py',
+    'scripts/run_backend_tests.py',
+    'scripts/run_lighthouse_tests.py',
+    'scripts/run_portserver.py',
+    'scripts/run_presubmit_checks.py',
+    'scripts/run_tests.py',
+    'scripts/start.py',
+    'setup.py',
+    'scripts/check_backend_associated_test_file.py',
+    'scripts/third_party_size_check.py',
+    'scripts/linters/warranted_angular_security_bypasses.py',
+    'core/feconf.py',
+    'core/domain/cron_services.py',
+    'core/platform/datastore/cloud_datastore_services.py',
+    'core/platform/transactions/cloud_transaction_services.py',
+]
+
+
+def main() -> None:
+    """Finds the non-empty backend files that lack an associated test file."""
+
+    all_backend_files = []
+    for (root, _, files) in os.walk(TOPMOST_LEVEL_PATH):
+        for file in files:
+            full_path = os.path.join(root, file)
+            _, file_extension = os.path.splitext(full_path)
+            if (file_extension == '.py') and (
+                '/third_party/' not in full_path) and (
+                    '/node_modules/' not in full_path):
+                all_backend_files.append(full_path)
+
+    all_backend_files.sort()
+    files_without_test = []
+    for file in all_backend_files:
+        if not file.endswith('_test.py'):
+            # Replace last occurance of '.py' with '_test.py' to get the
+            # associated test file.
+            to_be_replaced = '.py'
+            to_replace_with = '_test.py'
+            test_file = file[::-1].replace(
+                to_be_replaced[::-1], to_replace_with[::-1], 1)[::-1]
+            if test_file not in all_backend_files:
+                files_without_test.append(file)
+
+    non_empty_files = []
+    for file in files_without_test:
+        line_count = 0
+        with open(file, 'r', encoding='utf8') as f:
+            line_count = len(f.readlines())
+        if line_count > 0:
+            non_empty_files.append(file[2:])
+
+    errors = ''
+    for file in non_empty_files:
+        if file not in FILES_WITHOUT_ASSOCIATED_TEST_FILES:
+            errors += (
+                '\033[1m{}\033[0m does not have an associated test file.'
+                ' Make sure it\'s fully tested by its associated test'
+                ' file.\n'.format(file))
+
+    if errors:
+        print('-------------------------------------------')
+        print('Backend associated test file checks failed.')
+        print('-------------------------------------------')
+        logging.error(errors)
```

Comment:
Why is logging used here?

Reply:
I first collect all the error messages for the files which do not have an associated test file. Then after printing the banner `Backend associated test file check failed.`, I output all the errors at once using logging. Please take a look at **Failed pre-push** tab in **Proof that changes are correct** section.

Reply:
But why don't you just print them?

Reply:
`logging` allows us to print the error messages with various flags like error, warning, info etc. We use it here so that we can print the error messages with the `ERROR` flag. Other scripts like [check_frontend_test_coverage](https://github.com/oppia/oppia/blob/develop/scripts/check_frontend_test_coverage.py#L208) and [run_portserver](https://github.com/oppia/oppia/blob/develop/scripts/run_portserver.py#L518) also use the same syntax.

Reply:
Ok